core/src/main/scala/kafka/consumer/BaseConsumer.scala

```diff
@@ -91,6 +91,23 @@ class NewShinyConsumer(topic: Option[String], partitionId: Option[Int], offset:
     }
   }
 
+  private def seekToRealPositionsBeforeExit() {
+    val smallestUnconsumedOffsets = collection.mutable.Map[TopicPartition, Long]()
+    while (recordIter.hasNext) {
+      val record = recordIter.next()
+      val unconsumedPartition = new TopicPartition(record.topic(), record.partition())
+      // only insert the smallest offset for each partition
+      val currentOffset = smallestUnconsumedOffsets.getOrElse(unconsumedPartition, {
+        smallestUnconsumedOffsets += unconsumedPartition -> record.offset()
+        record.offset()
+      })
+      if (currentOffset > record.offset()) {
+        smallestUnconsumedOffsets += unconsumedPartition -> record.offset()
```

Comment:
@hachikuji Is the `if` block necessary?  Can we safely assert that the firstly encountered record for some partition must be of the smallest offset? If yes, we don't need to check the offset things. Instead we push that record's offset to the map. In my test env, it is indeed that the first read record per partition has the smallest offset. Does it make sense?

Reply:
Yes, I think you are right. If the map already has an entry, it must be the smallest.